### PR TITLE
Add more flags for image conversion (`--estargz-min-chunk-size`, etc.)

### LINKED ISSUE #1367: Make it possible to set the zstd compression level
### What is the problem you're trying to solve

I'm recompressing my images using `nerdctl image convert --oci --zstdchunked`. Unfortunately looking at the code it seems to use the default zstd compression level of 3, which is too low for my needs. The resulting zstd images are almost the same size as the gzipped ones.

### Describe the solution you'd like

Have a flag that allows setting the zstd compression level.

### Additional context

If I `nerdctl save` my images, get the layers out of the `.tar` generated by it, decompress them and recompress them with `zstd --ultra -22 -k` I almost halve the layer size for an acceptable compression time increase.

### PR DESCRIPTION
**Pull request by ktock**
Fixes  #1367

This adds the following flags and docs about the usage.

-  `--estargz-min-chunk-size=<SIZE>` : The minimal number of bytes of data must be written in one gzip stream (requires stargz-snapshotter >= v0.13.0). Useful for creating a smaller eStargz image (refer to [`./docs/stargz.md`](./docs/stargz.md) for details).
-  `--estargz-external-toc` : Separate TOC JSON into another image (called \"TOC image\"). The name of TOC image is the original + \"-esgztoc\" suffix. Both eStargz and the TOC image should be pushed to the same registry. (requires stargz-snapshotter >= v0.13.0) Useful for creating a smaller eStargz image (refer to [`./docs/stargz.md`](./docs/stargz.md) for details). :warning: This flag is experimental and subject to change.
-  `--estargz-keep-diff-id`: Convert to esgz without changing diffID (cannot be used in conjunction with '--estargz-record-in'. must be specified with '--estargz-external-toc')

### CODE REVIEW

**Comment by ktock** [comment]
Will rebase it soon after #1496 is merged. cc @AkihiroSuda

File: docs/experimental.md [hunk 1]
```diff
@@ -4,10 +4,8 @@ The following features are experimental and subject to change:
 
 - [Windows containers](https://github.com/containerd/nerdctl/issues/28)
 - [FreeBSD containers](./freebsd.md)
-- Importing an external eStargz record JSON file with `nerdctl image convert --estargz-record-in=FILE` .
+- Flags of `nerdctl image convert`: `--estargz-record-in=FILE` and `--zstdchunked-record-in=FILE` (Importing an external eStargz record JSON file), `--estargz-external-toc` (Separating TOC JSON to another image).
   eStargz itself is out of experimental.
-- Importing an external zstd record JSON file with `nerdctl image convert --zstdchunked-record-in=FILE` .
-  zstd itself is out of experimental.
```
Comment (AkihiroSuda) [inline docs/experimental.md:10]: Is zstd experimental?

File: docs/experimental.md [hunk 2]
```diff
@@ -4,10 +4,8 @@ The following features are experimental and subject to change:
 
 - [Windows containers](https://github.com/containerd/nerdctl/issues/28)
 - [FreeBSD containers](./freebsd.md)
-- Importing an external eStargz record JSON file with `nerdctl image convert --estargz-record-in=FILE` .
+- Flags of `nerdctl image convert`: `--estargz-record-in=FILE` and `--zstdchunked-record-in=FILE` (Importing an external eStargz record JSON file), `--estargz-external-toc` (Separating TOC JSON to another image).
```
Comment (AkihiroSuda) [inline docs/experimental.md:7]: Experimental flags need to check the `experimental` config

https://github.com/containerd/nerdctl/blob/b71443b2a3abc6288944c72229a2e21a1ae7ec6e/cmd/nerdctl/pull.go#L157-L164

File: docs/stargz.md [hunk 3]
```diff
@@ -89,3 +89,89 @@ $ nerdctl push example.com/foo:estargz
 NOTE: `--estargz` should be specified in conjunction with `--oci`
 
 Stargz Snapshotter is not needed for building stargz images.
+
+## Tips for image conversion
+
+### Tips 1: Using zstd instead of gzip (a.k.a. zstd:chunked)
+
+You can use zstd compression with lazy pulling support (a.k.a zstd:chunked) instead of gzip.
```
Comment (AkihiroSuda) [inline docs/stargz.md:97]: Pros and cons

File: docs/stargz.md [hunk 4]
```diff
@@ -89,3 +89,89 @@ $ nerdctl push example.com/foo:estargz
...
+
+
+```console
+$ nerdctl build -t example.com/foo .
+$ nerdctl image convert --zstdchunked --oci example.com/foo example.com/foo:zstdchunked
+$ nerdctl push example.com/foo:zstdchunked
+```
+
+### Tips 2: Creating smaller eStargz images
```
Comment (AkihiroSuda) [inline docs/stargz.md:106]: Tips 2 seems more important than tips 1, so they should be probably swapped

File: cmd/nerdctl/image_convert.go [hunk 5]
```diff
@@ -245,10 +249,76 @@ func imageConvertAction(cmd *cobra.Command, args []string) error {
...
+		if err != nil {
+			return err
+		}
+		is := client.ImageService()
+		_ = is.Delete(ctx, newI.Name)
+		finimg, err := is.Create(ctx, *newI)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintln(cmd.OutOrStdout(), "extra image:", finimg.Name)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/image_convert.go:268]: Is this string machine parsable?

Probably `--format=json` should be also available.
Can be another PR though.

File: docs/experimental.md (hunk 1 above)
Comment (ktock) [reply docs/experimental.md:10]: I think zstd isn't experimental.

File: docs/experimental.md (hunk 2 above)
Comment (ktock) [reply docs/experimental.md:8]: Fixed

File: docs/stargz.md (hunk 3 above)
Comment (ktock) [reply docs/stargz.md]: Thank you for the comment. Added them.

File: docs/stargz.md (hunk 4 above)
Comment (ktock) [reply docs/stargz.md]: Fixed.

File: cmd/nerdctl/image_convert.go (hunk 5 above)
Comment (ktock) [reply cmd/nerdctl/image_convert.go]: Added flag `--format=json` to print minimal information about the created images.

```console
# nerdctl image convert --format=json --oci --estargz --estargz-external-toc ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-ex
{
    "Image": {
        "Name": "registry2:5000/ubuntu:22.04-ex",
        "Digest": "sha256:3059dd5d9c404344e0b7c43d9782de8cae908531897262b7772103a0b585bbee"
    },
    "ExtraImages": [
        {
            "Name": "registry2:5000/ubuntu:22.04-ex-esgztoc",
            "Digest": "sha256:18c042b6eb8b208a8b97c88ad6d5186893141cb4ff43fab8c9bbb936eba8ae89"
        }
    ]
}
```

```console
# nerdctl image convert --format=json --oci --estargz ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-ex
{
    "Image": {
        "Name": "registry2:5000/ubuntu:22.04-ex",
        "Digest": "sha256:6d17b10f2d09267705f9500f07599cc2db57bee2c3175a1d0f91c0d88f43c9ec"
    },
    "ExtraImages": null
}
```

File: docs/experimental.md [hunk 6]
```diff
@@ -1,13 +1,12 @@
...
+See [`./config.md`](config.md) about how to enable these features.
 
 - [Windows containers](https://github.com/containerd/nerdctl/issues/28)
 - [FreeBSD containers](./freebsd.md)
-- Importing an external eStargz record JSON file with `nerdctl image convert --estargz-record-in=FILE` .
-  eStargz itself is out of experimental.
-- Importing an external zstd record JSON file with `nerdctl image convert --zstdchunked-record-in=FILE` .
-  zstd itself is out of experimental.
+- Flags of `nerdctl image convert`: `--estargz-record-in=FILE` and `--zstdchunked-record-in=FILE` (Importing an external eStargz record JSON file), `--estargz-external-toc` (Separating TOC JSON to another image).
+  eStargz and zstd itself are out of experimental.
```
Comment (AkihiroSuda) [inline docs/experimental.md:9]:
```suggestion
  eStargz and zstd themselves are out of experimental.
```

File: cmd/nerdctl/image_convert.go [hunk 7]
```diff
@@ -387,3 +484,34 @@ func imageConvertShellComplete(cmd *cobra.Command, args []string, toComplete str
 	// show image names
 	return shellCompleteImageNames(cmd)
 }
+
+type imageInfo struct {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/image_convert.go:488]: This name  will cause confusion when placed in the `main` package.
Should be moved to somewhere under `pkg`

File: cmd/nerdctl/image_convert.go [hunk 8]
```diff
@@ -387,3 +484,34 @@ func imageConvertShellComplete(cmd *cobra.Command, args []string, toComplete str
...
+	Name   string // image name
+	Digest string // image target digest (i.e. index digest or manifest digest)
+}
```
Comment (AkihiroSuda) [inline cmd/nerdctl/image_convert.go:491]: Needs `json:` tags

File: cmd/nerdctl/image_convert.go [hunk 9]
```diff
@@ -387,3 +484,34 @@ func imageConvertShellComplete(cmd *cobra.Command, args []string, toComplete str
 	// show image names
 	return shellCompleteImageNames(cmd)
 }
+
+type imageInfo struct {
+	Name   string // image name
+	Digest string // image target digest (i.e. index digest or manifest digest)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/image_convert.go:490]: Can we just name and digest into a single `Image` string concatenated with `@`?

Comment (ktock) [reply cmd/nerdctl/image_convert.go]: Thank you for the review. Fixed the struct format.

```console
# nerdctl image convert --format=json --oci --estargz --estargz-external-toc ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-ex
{
    "image": "registry2:5000/ubuntu:22.04-ex@sha256:3059dd5d9c404344e0b7c43d9782de8cae908531897262b7772103a0b585bbee",
    "extra_images": [
        "registry2:5000/ubuntu:22.04-ex-esgztoc@sha256:18c042b6eb8b208a8b97c88ad6d5186893141cb4ff43fab8c9bbb936eba8ae89"
    ]
}
# nerdctl image convert --oci --estargz --estargz-external-toc ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-ex
INFO[0006] Extra image(0) registry2:5000/ubuntu:22.04-ex-esgztoc 
sha256:3059dd5d9c404344e0b7c43d9782de8cae908531897262b7772103a0b585bbee
```

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Review by AkihiroSuda — APPROVED** [review]
Thanks
